src/assignment-5/CollateralizedVault.sol

```diff
@@ -0,0 +1,168 @@
+        if (getRequiredCollateral(borrows[msg.sender] + amount) > deposits[msg.sender]) {
+            revert NotEnoughCollateral();
+        }
+
+        borrows[msg.sender] += amount;
+
+        underlying.transfer(msg.sender, amount);
+    }
+
+    /// @notice Pays back an open debt
+    /// @param amount The amount of `underlying` token to pay back
+    function repay(uint256 amount) public {
+        borrows[msg.sender] -= amount;
+
+        underlying.transferFrom(msg.sender, address(this), amount);
+    }
+
+    /// @notice Withdraws part of the collateral
+    ///     Only allowed to withdraw as long as the collateral left is higher in value than the debt
+    /// @param collateralAmount The amount of `collateral` token to withdraw
+    function withdraw(uint256 collateralAmount) public {
+        uint256 requiredCollateral = getRequiredCollateral(borrows[msg.sender]);
+
+        if (deposits[msg.sender] - collateralAmount < requiredCollateral) {
+            revert TooMuchDebt();
+        }
+
+        deposits[msg.sender] -= collateralAmount;
+
+        collateral.transfer(msg.sender, collateralAmount);
+    }
+
+    /// @notice Admin: liquidate a user debt if the collateral value falls below the debt
+    /// @param user The user to liquidate
+    /// @dev Only admin is allowed to liquidate
+    function liquidate(address user) onlyOwner public {
+        uint256 requiredCollateral = getRequiredCollateral(borrows[user]);
+        if (deposits[user] >= requiredCollateral) {
+            revert UserDebtIsSufficientlyCollateralized();
+        }
+
+        delete borrows[user];
+        delete deposits[user];
+    }
+
+    /// @notice Returns the required amount of collateral in order to borrow `borrowAmount`
+    function getRequiredCollateral(uint256 borrowAmount) public view returns (uint256 requiredCollateral) {
+        requiredCollateral = mulup(borrowAmount, getPrice(), oracleDecimals);
+        requiredCollateral = scaleInteger(requiredCollateral, underlyingDecimals, collateralDecimals);
+    }
+
+    /// @dev Scales a fixed point interger from `fromDecimals` to `toDecimals`
+    function scaleInteger(uint256 x, uint8 fromDecimals, uint8 toDecimals) public pure returns (uint256) {
+        if (toDecimals >= fromDecimals) {
+            return x * 10**(toDecimals - fromDecimals);
+        } else {
+            return x / 10**(fromDecimals - toDecimals);
+        }
+    }
+
+    /// @notice Returns the price of the underlying token in collateral token units
+    function getPrice() public view returns (uint) {
+        (
+            /*uint80 roundID*/,
+            int price,
+            /*uint startedAt*/,
+            /*uint timeStamp*/,
+            /*uint80 answeredInRound*/
+        ) = oracle.latestRoundData();
+        return toUint256(price);
+    }
+
+    function mulup(uint256 x, uint256 y, uint8 decimals) internal pure returns (uint256 z) {
+        z = (x * y + 10**decimals - 1) / 10**decimals;
```

Comment:
wonder how this formula works.. usually we multiply by 10^18 and then divide.. but here we have plus operator.. because of order.. (((x*y) + (10^18) - 1) / 10^18) ??

Reply:
so goal is to multiply and round up.

the formula is equal to `(x*y)/1e18 + (1e18 - 1)/1e18`
so if `x*y` is not divisible by `1e18`, adding  `(1e18 - 1)/1e18` will roll it over to the next integer, and then the implicit round down will happen. so effectively, it would be round up

Reply:
got it.. makes sense when I look at "10**decimals - 1" as a constant..